File: apps/mobile/src/modules/photos/dateRange.ts
import type { GalleryPhoto } from '@/modules/galleries/types'

const monthFormatters = new Map<string, Intl.DateTimeFormat>()

function getMonthFormatter(locale: string, includeYear: boolean): Intl.DateTimeFormat {
  const key = `${locale}:${includeYear ? 'year' : 'current'}`
  const existing = monthFormatters.get(key)
  if (existing) {
    return existing
  }
  const formatter = new Intl.DateTimeFormat(locale, {
    month: 'long',
    ...(includeYear ? { year: 'numeric' } : {}),
  })
  monthFormatters.set(key, formatter)
  return formatter
}

function parseDate(value: string | null): Date | null {
  if (!value) {
    return null
  }
  const date = new Date(value)
  return Number.isNaN(date.getTime()) ? null : date
}

export function formatVisibleMonthAnchor(
  photos: GalleryPhoto[],
  startIndex: number,
  endIndex: number,
  locale: string,
): string | null {
  if (photos.length === 0) {
    return null
  }
  const start = Math.max(0, Math.min(startIndex, photos.length - 1))
  const end = Math.max(start, Math.min(endIndex, photos.length - 1))

  const counts = new Map<number, number>()
  for (let i = start; i <= end; i++) {
    const date = parseDate(photos[i].dateTaken)
    if (!date) {
      continue
    }
    const bucket = date.getFullYear() * 12 + date.getMonth()
    counts.set(bucket, (counts.get(bucket) ?? 0) + 1)
  }
  if (counts.size === 0) {
    return null
  }

  let anchor = -1
  let anchorCount = 0
  for (const [bucket, count] of counts) {
    if (count > anchorCount || (count === anchorCount && bucket > anchor)) {
      anchor = bucket
      anchorCount = count
    }
  }

  const year = Math.floor(anchor / 12)
  const month = anchor % 12
  const includeYear = year !== new Date().getFullYear()
  return getMonthFormatter(locale, includeYear).format(new Date(year, month, 1))
}

File: apps/mobile/modules/photo-masonry/ios/Models/DateRange.swift
import Foundation

enum DateRange {
  static func visibleMonthAnchor(
    photos: [GalleryPhoto],
    startIndex: Int,
    endIndex: Int,
    localeIdentifier: String,
    now: Date = .now,
    timeZone: TimeZone = .current
  ) -> String? {
    guard !photos.isEmpty else { return nil }
    let start = max(0, min(startIndex, photos.count - 1))
    let end = max(start, min(endIndex, photos.count - 1))
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = timeZone
    var counts: [Int: Int] = [:]
    for photo in photos[start...end] {
      guard let value = photo.dateTaken,
            let date = PhotoDateParser.date(value, timeZone: timeZone)
      else { continue }
      let year = calendar.component(.year, from: date)
      let month = calendar.component(.month, from: date) - 1
      counts[year * 12 + month, default: 0] += 1
    }
    guard let anchor = counts.keys.max(by: { lhs, rhs in
      let leftCount = counts[lhs] ?? 0
      let rightCount = counts[rhs] ?? 0
      return leftCount == rightCount ? lhs < rhs : leftCount < rightCount
    }) else { return nil }
    let year = anchor / 12
    let month = anchor % 12 + 1
    var components = DateComponents()
    components.calendar = calendar
    components.timeZone = timeZone
    components.year = year
    components.month = month
    components.day = 1
    guard let date = components.date else { return nil }
    let formatter = DateFormatter()
    formatter.locale = Locale(identifier: localeIdentifier)
    formatter.timeZone = timeZone
    let includeYear = year != calendar.component(.year, from: now)
    formatter.setLocalizedDateFormatFromTemplate(includeYear ? "yMMMM" : "MMMM")
    return formatter.string(from: date)
  }
}
